File: YtDlp.fs
module Archivist.YtDlp

open System
open System.IO
open System.Text.Json
open System.Threading.Tasks
open Archivist.Domain
open Archivist.Paths
open Archivist.ProcessRunner

let executableName = "yt-dlp"

let private stringOptionOfNullable (value: string | null) =
    if isNull value || String.IsNullOrWhiteSpace value then None else Some value

let private tryGetString (name: string) (json: JsonElement) =
    match json.TryGetProperty name with
    | true, value when value.ValueKind = JsonValueKind.String ->
        value.GetString() |> stringOptionOfNullable
    | _ ->
        None

let probeArgs (url: string) = [ "--dump-json"; "--skip-download"; "--playlist-end"; string 1; url ]

let probe (url: string) : Task<ProbeOutcome> =
    task {
        let args = probeArgs url
        let! result = run executableName args

        if result.exitCode <> 0 then
            return ProbeFailed result.stderr
        else
            try
                use doc = JsonDocument.Parse(result.stdout)
                let root = doc.RootElement

                return
                    ProbeSuccess
                        { channel = tryGetString "channel" root
                          channelId = tryGetString "channel_id" root
                          channelHandle = tryGetString "channel_handle" root
                          uploader = tryGetString "uploader" root
                          uploaderId = tryGetString "uploader_id" root }
            with ex ->
                return ProbeFailed $"Failed to parse yt-dlp metadata: {ex.Message}"
    }

let private pathCombineForYoutube left right =
    if String.IsNullOrWhiteSpace left then right
    elif String.IsNullOrWhiteSpace right then left
    else Path.Combine(left, right)

let private outputTemplate (config: Config) (target: Target) =
    match target.outputTemplate with
    | Some template when not (String.IsNullOrWhiteSpace template) -> template
    | _ -> config.defaultYoutubeTemplate

let private targetUrls (target: Target) =
    match target.urls with
    | Some urls ->
        urls
        |> List.map (fun url -> url.Trim())
        |> List.filter (fun url -> not (String.IsNullOrWhiteSpace url))
        |> function
            | [] -> [ target.url ]
            | urls -> urls
    | None -> [ target.url ]

let buildSyncArgs (config: Config) (label: string) (target: Target) =
    let archivePath = youtubeArchiveFile config label

    [ "--download-archive"
      archivePath
      "--paths"
      target.subdir
      |> Option.defaultValue ""
        |> fun subdir -> pathCombineForYoutube config.youtubeDir subdir
      "-o"
      outputTemplate config target ]
    @ targetUrls target

let sync (config: Config) (label: string) (target: Target) : Task<ProcessResult> =
    task {
        Directory.CreateDirectory(youtubeArchiveDirectory config) |> ignore
        let args = buildSyncArgs config label target
        return! run executableName args
    }

File: src/yt_dlp.rs
use crate::config::config_option_args;
use crate::paths::youtube_archive_file;
use crate::process::run_process;
use crate::types::{Config, ProbeInfo, SourceType, Target};
use std::path::Path;

pub fn probe_args(url: &str) -> Vec<String> {
    vec![
        "--dump-json".to_string(),
        "--skip-download".to_string(),
        "--playlist-end".to_string(),
        "1".to_string(),
        url.to_string(),
    ]
}

pub fn probe(url: &str) -> Result<ProbeInfo, String> {
    let result = run_process("yt-dlp", &probe_args(url)).map_err(|error| error.to_string())?;
    if result.exit_code != 0 {
        return Err(result.stderr);
    }
    serde_json::from_str(&result.stdout)
        .map_err(|error| format!("Failed to parse yt-dlp metadata: {error}"))
}

pub fn build_sync_args(
    config: &Config,
    label: &str,
    target: &Target,
) -> Result<Vec<String>, String> {
    let mut args = config_option_args(&config.yt_dlp_options, "yt_dlp_options")?;
    args.extend([
        "--download-archive".to_string(),
        youtube_archive_file(config, label)
            .to_string_lossy()
            .into_owned(),
        "--paths".to_string(),
        if target.subdir {
            Path::new(&config.youtube_dir)
                .join(label)
                .to_string_lossy()
                .into_owned()
        } else {
            config.youtube_dir.clone()
        },
        "-o".to_string(),
        target
            .output_template
            .as_deref()
            .filter(|template| !template.trim().is_empty())
            .unwrap_or(&config.default_youtube_template)
            .to_string(),
    ]);
    args.extend(target.urls.clone());
    Ok(args)
}

pub fn expand_playlist_urls<F>(
    urls: &mut Vec<String>,
    source_type: SourceType,
    include_all: bool,
    mut confirm: F,
) where
    F: FnMut(&str) -> bool,
{
    if source_type != SourceType::YouTube {
        return;
    }

    let base_urls = urls
        .iter()
        .filter_map(|url| playlist_base_url(url))
        .collect::<Vec<_>>();

    for base_url in base_urls {
        if urls.iter().any(|url| same_url(url, &base_url)) {
            continue;
        }

        if include_all || confirm(&base_url) {
            urls.push(base_url);
        }
    }
}

pub fn playlist_base_url(url: &str) -> Option<String> {
    let trimmed = url.trim().trim_end_matches('/');
    let lower = trimmed.to_ascii_lowercase();
    let suffix = "/playlists";

    if !(lower.contains("youtube.com") || lower.contains("youtu.be")) || !lower.ends_with(suffix) {
        return None;
    }

    Some(trimmed[..trimmed.len() - suffix.len()].to_string())
}

fn same_url(left: &str, right: &str) -> bool {
    left.trim()
        .trim_end_matches('/')
        .eq_ignore_ascii_case(right.trim().trim_end_matches('/'))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ytdlp_options_are_prepended_to_sync_args() {
        let config = Config {
            yt_dlp_options: Some(toml::Value::Array(vec![
                toml::Value::String("--ignore-errors".to_string()),
                toml::Value::String("--no-warnings".to_string()),
            ])),
            ..Config::default()
        };
        let target = Target {
            urls: vec!["https://www.youtube.com/example".to_string()],
            source: "youtube".to_string(),
            subdir: false,
            output_template: None,
        };

        let args = build_sync_args(&config, "example", &target).expect("valid options");

        assert_eq!(args[0], "--ignore-errors");
        assert_eq!(args[1], "--no-warnings");
        assert_eq!(args[2], "--download-archive");
    }

    #[test]
    fn youtube_playlist_base_url_strips_playlists_suffix() {
        let base = playlist_base_url("https://www.youtube.com/@example/playlists/");

        assert_eq!(base.as_deref(), Some("https://www.youtube.com/@example"));
    }

    #[test]
    fn include_all_adds_youtube_playlist_base_url_once() {
        let mut urls = vec![
            "https://www.youtube.com/@example/playlists".to_string(),
            "https://www.youtube.com/@example/".to_string(),
        ];

        expand_playlist_urls(&mut urls, SourceType::YouTube, true, |_| false);

        assert_eq!(urls.len(), 2);
    }
}
